[DevTools] Use Popover API for TraceUpdates highlighting

yongsk0066 opened the pull request:
## Summary

When using React DevTools to highlight component updates, the highlights would sometimes appear behind elements that use the browser's [top-layer](https://developer.mozilla.org/en-US/docs/Glossary/Top_layer) (such as `<dialog>` elements or components using the Popover API). This made it difficult to see which components were updating when they were inside or behind top-layer elements.

  This PR fixes the issue by using the Popover API to ensure that highlighting appears on top of all content, including elements in the top-layer. The implementation maintains backward compatibility with browsers that don't support the Popover API.

## How did you test this change?

I tested this change in the following ways:

1. Manually tested in Chrome (which supports the Popover API) with:
   - Created a test application with React components inside `<dialog>` elements and custom elements using the Popover API
   - Verified that component highlighting appears above these elements when they update
   - Confirmed that highlighting displays correctly for nested components within top-layer elements

2. Verified backward compatibility:
   - Tested in browsers without Popover API support to ensure fallback behavior works correctly
   - Confirmed that no errors occur and highlighting still functions as before

3. Ran the React DevTools test suite:
   - All tests pass successfully
   - No regressions were introduced

[demo-page](https://devtools-toplayer-demo.vercel.app/)
[demo-repo](https://github.com/yongsk0066/devtools-toplayer-demo)

### AS-IS
https://github.com/user-attachments/assets/dc2e1281-969f-4f61-82c3-480153916969

### TO-BE
https://github.com/user-attachments/assets/dd52ce35-816c-42f0-819b-0d5d0a8a21e5

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -204,6 +216,7 @@ export function destroy(agent: Agent): void {
 
 function initialize(): void {
   canvas = window.document.createElement('canvas');
+  canvas.setAttribute('popover', 'auto');
```
Can this popover be dismissed automatically by the user? By hitting `Esc`, for example?

yongsk0066 replied:
Hmm, you're right. I missed that. I've switched it to manual mode and handled the elements that might interfere with the user. Thank you!

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -45,6 +45,12 @@ function drawNative(nodeToData: Map<HostInstance, Data>, agent: Agent) {
   if (canvas === null) {
     initialize();
+  } else {
+    try {
+      if (canvas.hasAttribute('popover') && !canvas.matches(':popover-open')) {
+        canvas.showPopover();
+      }
```
Are you gating it with try/catch because for feature checking?

We might bump the minimum required version for Chrome / other browsers here:
https://github.com/facebook/react/blob/main/packages/react-devtools-extensions/chrome/manifest.json#L7

So you won't need to worry about if its supported or not

yongsk0066 replied:
I've added Edge as well! Yes, Firefox requires 125 or higher, but it's already set to 128, so we're good.

hoxyq commented:
Please see the failed Flow checks, overall the changes are seemed necessary, thanks for fixing this

yongsk0066 commented:
@hoxyq Thanks for the review! i've fixed all the Flow errors and updated the implementation. I've added appropriate $FlowFixMe annotations since Flow doesn't recognize the Popover API yet. All tests are now passing!

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -204,6 +234,9 @@ export function destroy(agent: Agent): void {
 
 function initialize(): void {
   canvas = window.document.createElement('canvas');
+  // canvas.setAttribute('popover', 'manual');
```
Should this be uncommented?

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -43,8 +43,30 @@ function drawNative(nodeToData: Map<HostInstance, Data>, agent: Agent) {
+  // if there are no nodes to draw, detach from top layer
+  if (nodeToData.size === 0) {
+    if (canvas !== null) {
+      try {
```
Do you actually need to wrap it in try / catch now?

yongsk0066 replied:
Oh, I initially added that to solve Flow and type issues, but it became unnecessary after adding the annotations. I've removed it.

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -43,8 +43,30 @@ function drawNative(nodeToData: Map<HostInstance, Data>, agent: Agent) {
+        if (canvas.matches(':popover-open')) {
+          // $FlowFixMe[prop-missing]
+          // $FlowFixMe[incompatible-use]
+          canvas.hidePopover();
```
Why it needs to be hidden in this case?

yongsk0066 replied:
If we don't remove it from the toplayer, it would be pushed behind when other toplayer elements appear later.

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -213,9 +246,18 @@ function initialize(): void {
   root.insertBefore(canvas, root.firstChild);
+
+  try {
+    // $FlowFixMe[prop-missing]
+    // $FlowFixMe[incompatible-use]
```
Please specify every new FlowFixMe where Popover API is referenced by adding "Flow doesn't recognize Popover API" comment

yongsk0066 replied:
Thanks I've added it to all the necessary places!

yongsk0066 commented:
Hi @hoxyq, just wanted to check in to see if there's anything else needed from my side. 
Let me know if I can clarify or help further. Thanks!

By the way, I've also added Popover-related type definitions to [flow-typed](https://github.com/flow-typed/flow-typed/pull/4607).
However, since the version of Flow used in the React repository is still behind, 
we can't apply them here immediately. Perhaps we can in the future once it's updated.

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -43,8 +43,28 @@ function drawNative(nodeToData: Map<HostInstance, Data>, agent: Agent) {
   if (canvas === null) {
     initialize();
+  } else {
+    try {
```
Same as above, you don't need it to wrap it in try / catch, since we bumped minimum versions for the extension

Please also check other calls below

yongsk0066 replied:
Thanks for pointing this out! I've removed all try/catch blocks around Popover API calls.

hoxyq reviewed (commented):
Thanks for updating flow-typed as well!

Flow team does upgrades frequently, so it might be updated soon, but shouldn't block you from landing this change.

Overall looks good to me, see my comment, I will try to manually test this tomorrow.

hoxyq commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -191,7 +209,13 @@ function destroyNative(agent: Agent) {
   if (canvas !== null) {
+    // $FlowFixMe[prop-missing]: Flow doesn't recognize Popover API
+    // $FlowFixMe[incompatible-use]: Flow doesn't recognize Popover API
+    canvas.hidePopover();
```
I think you should check here if popover is opened to avoid getting InvalidStateException

yongsk0066 commented on packages/react-devtools-shared/src/backend/views/TraceUpdates/canvas.js:
```diff
@@ -43,8 +43,26 @@ function drawNative(nodeToData: Map<HostInstance, Data>, agent: Agent) {
+      }
+    }
+    return;
+  }
```
Ah, I see. Totally agree that we should first finish all `drawWeb` logic and then decide at the end whether or not to promote it to the top-layer. 
I think that approach better fits the responsibilities of each method. Thanks for clarifying this!
Considering your comments, I've tested two different approaches:

### 1. Hiding the canvas only when there's nothing to draw (the current method):
```javascript
function drawWeb(nodeToData: Map<HostInstance, Data>) {
  // ... previous code
  if (canvas !== null) {
    if (nodeToData.size === 0 && canvas.matches(':popover-open')) {
      canvas.hidePopover();
    } else if (nodeToData.size > 0 && !canvas.matches(':popover-open')) {
      canvas.showPopover();
    }
  }
} // end of drawWeb
```
Result: This method avoids frequent updates to the top-layer.
However, as you mentioned in your comment, if another top-layer element appears while highlights still remain, the highlight canvas might fall behind the new element.

### 2. Re-promoting canvas to top-layer on every update call:
```javascript
function drawWeb(nodeToData: Map<HostInstance, Data>) {
  // ... previous code
  if (canvas !== null) {
    if (nodeToData.size === 0 && canvas.matches(':popover-open')) {
      canvas.hidePopover();
      return;
    }
    if (canvas.matches(':popover-open')) {
      canvas.hidePopover();
    }
    canvas.showPopover();
  }
} // end of drawWeb
```
Result: Frequent changes to the top-layer occur, but the highlights consistently remain on top, ensuring they never get hidden by other elements.
I think method `#2` seems preferable in terms of clearly maintaining highlight visibility too. 
However, I couldn't find any mentions of performance issues in either the W3C review or blink-dev discussions regarding frequent `top-layer` changes. Unless highlights are triggered continuously at extremely high frequency over a prolonged period, there shouldn't be any noticeable performance impact.

yongsk0066 commented:
Thanks again for the review, @hoxyq !  
I've added the test to `react-devtools-shell` as suggested.

yongsk0066 commented:
How have you been, @hoxyq ?

It's been about a month, so I'm just leaving a comment to check if there's any additional feedback. I've been using this version for about a month without any major issues. No rush at all, just checking in. Thanks!

hoxyq commented:
> How have you been, @hoxyq ?
> 
> It's been about a month, so I'm just leaving a comment to check if there's any additional feedback. I've been using this version for about a month without any major issues. No rush at all, just checking in. Thanks!

Hey @yongsk0066, apologies for the delay. Don't worry, this will be in the next release of React DevTools, I was just prioritising other things. I have now finish other things and planning to get this tested next week.

Thank you for contributing to React DevTools.

hoxyq reviewed (approved):
Sorry for the delay, looks good. Thanks again!

https://github.com/facebook/react/pull/32703 is in my queue as well, I will also review and test it, probably worth rebasing it on `main`, once I merge this one.
